File: CSharp/MyDynamicList.cs
using System.Collections;

namespace DataStructures;

public class MyDynamicList 
{
    public const int DEFAULT_CAPACITY = 4;
}

public class MyDynamicList<T> : MyDynamicList , IList<T>
{
    public int Count { get; private set; }
    public int Capacity => _array.Length;
    public bool IsReadOnly => false;

    private T?[] _array;

    public MyDynamicList(T[] array)
    {
        var size = array.Length;
        _array = new T[size];
        Array.Copy(array,0,_array,0, size);
        Count = size;
    }

    public MyDynamicList()
    {
        _array = new T[DEFAULT_CAPACITY];
        Count = 0;
    }

    public MyDynamicList(int capacity)
    {
        _array = new T[capacity];
        Count = 0;
    }

    public T this[int index]
    {
        get
        {
            ValidateInBounds(index);
            return _array[index]!;
        }
        set
        {
            ValidateInBounds(index);
            _array[index] = value;
        }
    }

    public IEnumerator<T> GetEnumerator()
    {
        for (var i = 0; i < Count; i++)
        {
            yield return _array[i]!;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public void Add(T item)
    {
        ++Count;
        EnsureCapacity();
        _array[Count - 1] = item;
    }

    private void EnsureCapacity()
    {
        if (Count <= Capacity) return;

        var newCapacity = Capacity * 2; 
        if (newCapacity < 1)
        {
            newCapacity = 1;
        }

        var newArray = new T?[newCapacity];
        for (var i = 0; i < _array.Length; i++)
        {
            newArray[i] = _array[i];
        }
        _array = newArray;
    }

    public void Clear()
    {
        for (var index = 0; index < _array.Length; index++)
        {
            _array[index] = default;
        }
        Count = 0;
    }

    public bool Contains(T item)
    {
        return IndexOf(item) != -1;
    }

    public void CopyTo(T[] array, int arrayIndex)
    {
        for (var index = arrayIndex; index < Count; index++)
        {
            array[index] = _array[index]!;
        }
    }

    public bool Remove(T item)
    {
        var index = IndexOf(item);
        if (index == -1) return false;
        RemoveAt(index);
        return true;
    }

    public int IndexOf(T item)
    {
        for (var i = 0; i < Count; i++)
        {
            if (Equals(_array[i], item)) return i;
        }
        return -1;
    }

    public void Insert(int index, T item)
    {
        Count++;
        EnsureCapacity();
        for (var i = Count - 2; i >= index; i--)
        {
            _array[i + 1] = _array[i];
        }
        _array[index] = item;
    }

    public void RemoveAt(int index)
    {
        RemoveAll((_, i) => i == index);
    }

    public void RemoveAll(Func<T,bool> predicate)
    {
        RemoveAll((item,_) => predicate(item));
    }

    public void RemoveAll(Func<T,int, bool> predicate)
    {
        var left = -1;
        var newCount = 0;
        for (var right = 0; right < Count; right++)
        {
            if (predicate(_array[right]!, right))
            {
                _array[right] = default;
                continue;
            }
            left++;
            newCount++;
            var temp = _array[left];
            _array[left] = _array[right];
            _array[right] = temp;
        }
        Count = newCount;
    }

    public void ValidateInBounds(int index)
    {
        if (0 > index || index >= Count)
        {
            throw new IndexOutOfRangeException();
        }
    }

}

File: Cpp/MyDynamicList.cpp
#include "pch.h"
#include <exception>
namespace ds
{
	constexpr int default_capacity = 4;

	MyDynamicList::Iterator::Iterator(int* pointer): _pointer(pointer) { }

	//iterator
	MyDynamicList::Iterator& MyDynamicList::Iterator::operator++()
	{
		_pointer++;
		return *this;
	}

	MyDynamicList::Iterator MyDynamicList::Iterator::operator++(int)
	{
		const Iterator self = *this;
		++(*this);
		return self;
	}

	MyDynamicList::Iterator& MyDynamicList::Iterator::operator--()
	{
		_pointer--;
		return *this;
	}

	MyDynamicList::Iterator MyDynamicList::Iterator::operator--(int)
	{
		const Iterator self = *this;
		--(*this);
		return self;
	}

	int MyDynamicList::Iterator::operator*() const
	{
		return *_pointer;
	}

	int* MyDynamicList::Iterator::operator->() const
	{
		return _pointer;
	}

	MyDynamicList::Iterator::operator int*() const
	{
		return _pointer;
	}

	bool MyDynamicList::Iterator::operator==(const Iterator& other) const
	{
		return _pointer == other._pointer;
	}

	bool MyDynamicList::Iterator::operator!=(const Iterator& other) const
	{
		return !(_pointer == other._pointer);
	}

	//constructors
	MyDynamicList::MyDynamicList() : MyDynamicList(default_capacity) { }

	MyDynamicList::MyDynamicList(MyDynamicList& copy)
	{
		array_ = new int[copy.get_capacity()];
		capacity_ = copy.get_capacity();
		for(const int i : copy)
		{
			add(i);
		}
	}

	MyDynamicList::MyDynamicList(const int capacity)
	{
		if (capacity <= 0) throw std::exception("Capacity can't be 0");
		array_ = new int[capacity];
		capacity_ = capacity;
	}

	MyDynamicList::MyDynamicList(int array[], int size_of_array)
	{
		array_ = new int[size_of_array];
		for(int i = 0; i < size_of_array; i++)
		{
			array_[i] = array[i];
		}
		count_ = size_of_array;
		capacity_ = size_of_array;
	}

	//public functions
	void MyDynamicList::add(const int value)
	{
		++count_;
		ensure_capacity();
		array_[count_ - 1] = value;
	}

	void MyDynamicList::remove_at(int index)
	{
		validate_in_bounds(index);

		const std::function<bool(int, int)> func = 
			[index](const int _, const int i)
		{
			return index == i;
		};

		remove_all(func);
	}

	void MyDynamicList::remove_all(const std::function<bool(int)>& predicate)
	{
		const std::function<bool(int, int)> func = [predicate](const int item, const int _)
		{
			return predicate(item);
		};

		remove_all(func);
	}

	void MyDynamicList::remove_all(const std::function<bool(int, int)>& predicate)
	{
		int left = -1;
		int new_count = 0;
		for (int right = 0; right < capacity_; right++)
		{
			const int value = *(array_ + right);

			if (predicate(value, right)) continue;
			
			left++;
			new_count++;

			const int temp = array_[left];
			array_[left] = array_[right];
			array_[right] = temp;
		}
		count_ = new_count;
	}

	int& MyDynamicList::operator[](const int index) const
	{
		validate_in_bounds(index);
		return *(array_ + index);
	}

	int MyDynamicList::get_count() const
	{
		return count_;
	}

	int MyDynamicList::get_capacity() const
	{
		return capacity_;
	}

	int MyDynamicList::get_at(const int index) const
	{
		validate_in_bounds(index);
		return *(array_ + index);
	}

	void MyDynamicList::set_at(const int index, const int value) const
	{

		validate_in_bounds(index);
		int* ref = array_ + index;
		*ref = value;
	}

	MyDynamicList::Iterator MyDynamicList::begin()
	{
		return Iterator(array_);
	}

	MyDynamicList::Iterator MyDynamicList::end()
	{
		return Iterator(array_ + count_);
	}

	MyDynamicList::~MyDynamicList()
	{
		delete[] array_;
	}

	//private functions
	void MyDynamicList::ensure_capacity()
	{
		if (count_ <= capacity_) return;

		const int new_capacity = capacity_ * 2;
		const auto copy = new int[new_capacity];

		for (int i = 0; i < capacity_; i++)
		{
			const int value = *(array_ + i);
			copy[i] = value;
		}

		delete[] array_;

		array_ = copy;
		capacity_ = new_capacity;
	}

	inline void MyDynamicList::validate_in_bounds(const int index) const
	{
		if (0 > index || index >= get_count())
		{
			throw std::exception("index out of bounds");
		}
	}
}
